OrbitVulkanLayer/DispatchTable.h

```diff
@@ -0,0 +1,413 @@
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdWriteTimestamp != nullptr);
+    return device_dispatch_table_.at(key).CmdWriteTimestamp;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdBeginDebugUtilsLabelEXT CmdBeginDebugUtilsLabelEXT(
+      const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdBeginDebugUtilsLabelEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdBeginDebugUtilsLabelEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdEndDebugUtilsLabelEXT CmdEndDebugUtilsLabelEXT(
+      const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdEndDebugUtilsLabelEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdEndDebugUtilsLabelEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdDebugMarkerBeginEXT CmdDebugMarkerBeginEXT(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdDebugMarkerBeginEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdDebugMarkerBeginEXT;
+  }
+
+  template <typename DispatchableType>
+  PFN_vkCmdDebugMarkerEndEXT CmdDebugMarkerEndEXT(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_dispatch_table_.contains(key));
+    CHECK(device_dispatch_table_.at(key).CmdDebugMarkerEndEXT != nullptr);
+    return device_dispatch_table_.at(key).CmdDebugMarkerEndEXT;
+  }
+
+  template <typename DispatchableType>
+  bool IsDebugMarkerExtensionSupported(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_supports_debug_marker_extension_.contains(key));
+    return device_supports_debug_marker_extension_.at(key);
+  }
+
+  template <typename DispatchableType>
+  bool IsDebugUtilsExtensionSupported(const DispatchableType& dispatchable_object) {
+    absl::ReaderMutexLock lock(&mutex_);
+    void* key = GetDispatchTableKey(dispatchable_object);
+    CHECK(device_supports_debug_utils_extension_.contains(key));
+    return device_supports_debug_utils_extension_.at(key);
+  }
+
+ private:
+  // In vulkan, every "dispatchable type" has as a very first field in memory a pointer to the
```

Comment:
Could you elaborate more on what a dispatchable type is? Just a device or instance? Or more? Is a command buffer that you mention in this comment also a dispatchable type?

Reply:
I expanded the comment here a bit. Hope it gets clearer now.